Introducing Keyframe-like animation definition schema (#2195)
software-mansion/react-native-reanimated: docs/docs/api/LayoutAnimations/KeyframeAnimations.md

---
id: keyframeAnimations
title: Keyframe Animations
sidebar_label: Keyframe Animations
---

The document explains how you can define complex animation using simple and popular animation definitions schema - Keyframes.

## How to define Keyframe animation?

### 1. Import Keyframe

```js
import { Keyframe } from 'react-native-reanimated';
```

### 2. Create Keyframe object, define initial and final state

In Keyframe's constructor pass object with definitions of your animation. Object keys correspond to animation progress,
so to 0 assign the style, you want for your object at the beginning of the animation and to 100 assign the style you want for your object to have at the end of the animation.

```js
import { Keyframe } from 'react-native-reanimated';

const keyframe = new Keyframe({
    0: {
      transform: [ { rotate: '0deg' }],
    },
    100: {

>       transform: [ { rotate: '45deg' }],

Reviewer: 🌵 Unnecessary whitespace in transform array.

Reviewer: Actually in most of the examples whitespace is inconsistent. Could you fix that, please?

    },
  }
```

### 3. Add middle points

Between edge points, you can define middle points in which you want your object to have certain style properties.

> Remember that for every keyframe you should provide the same set of style properties!

Reviewer: What happens if you omit them? I think better UX would be to preserve them with previous ones so using keyframes like:
```js
const keyframe = new Keyframe({
    0: {
      transform: [{ rotate: '0deg' }],
    },
    45: {
        transform: [{ transformX: 20 }]
    },
    100: {
      transform: [{ rotate: '45deg' }],
    },
  }
```

would keep rotation at 0 degrees on 45 key and add `transformX` that will be present on the last frame.

Author: It requires some changes in the parsing logic. I think preserving values with previous ones could be done quite easily. However, it will be harder to start animate new property in the middle of animation but I'm gonna think about it.

Reviewer: Starting from a default value for a property is the one way to go here. It's easy for most props by using zero values with few exceptions like color or opacity. I think this is quite intuitive behavior. You'll need to traverse values first and merge a zero keyframe with props that are specified later on.

\cc @piaskowyk

Author: I had a similar idea. What if we require to pass all used props to zero keyframe as it contains initial values? That way we don't have to handle those edge cases with color, opacity, etc.

Reviewer 2: I like this idea! We forgot about this before, but this approach is much better and more intuitive for users. Could we also add from-to syntax?
```js
const keyframe = new Keyframe({
  from: {
    transform: [{ rotate: '0deg' }],
  },
  to: {
    transform: [{ rotate: '45deg' }],
  },
});
```

Reviewer: > What if we require to pass all used props to zero keyframe as it contains initial values? 

This is a great solution and we can relax this restriction in the future if we want to.

    },
    45: {
        transform: [{ rotate: '100deg' }]
    },
    100: {
      transform: [ { rotate: '45deg' }],
    },
  }
```

### 4. Customize transitions using an easing function

```js
import { Keyframe, Easing } from 'react-native-reanimated';

const keyframe = new Keyframe({
    0: {
      transform: [ { rotate: '0deg' }],
    },
    45: {
        transform: [{ rotate: '100deg' }],
        easing: Easing.exp,
    },
    100: {
      transform: [ { rotate: '45deg' }],
    },
  }
```

## How to use keyframe animations?

Currently, you can define animations using keyframes only for entry and exit animations. 

### 1. Choose Animated Component which entering or exiting you want to animate
```js
    // AnimatedComponent - component created by createAnimatedComponent or imported from Reanimated
    // keyframe - Keyframe object
    <AnimatedComponent exiting={keyframe} >
```
### 2. Customize the animation
```js
    <AnimatedComponent exiting={keyframe.duration(3000).delay(200)} >
```
### 3. Make sure that your animated component is under an AnimatedLayout. If it's not then add AnimatedLayout somewhere above the component.
```js
    <AnimatedLayout> // +

>         <View> sth </View>

Reviewer: Please use `<Text />` if you want to show the rendering component with text. Adding text between any other component is invalid in RN and will error.

        <View> 

>             <AnimatedComponent exiting={keyframe.duration(3000).delay(200)} >

Reviewer 2: Could you add the missing slash before closing the bracket? `<AnimatedComponent />` This is in more than just here but I don't want to create too many comments.

>         </View>

Reviewer: What's the role of the empty `<View />` here?

    </AnimatedLayout> // +
```

## Available modifiers

> The order of Modifiers doesn't matter.

Reviewer: ```suggestion
The order of modifiers doesn't matter.
```

### duration
default: 300

> How long the animation should last

Reviewer: ```suggestion
How long the animation should last.
```

  }
```

> Instead of using '0' and '100', you can define edge points using `from` and `to` keywords. The result will be the same.

Reviewer: To be consistent with the numbers above.

```suggestion
Instead of using `0` and `100`, you can define edge points using `from` and `to` keywords. The result will be the same.
```
